core/chains/evm/txmgr/nonce_tracker.go

```diff
@@ -0,0 +1,201 @@
+package txmgr
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"slices"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/jpillora/backoff"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+	evmtypes "github.com/smartcontractkit/chainlink/v2/core/chains/evm/types"
+)
+
+type NonceTrackerTxStore interface {
+	GetAllTransactions(context.Context, common.Address, *big.Int) (txs []Tx, err error)
+}
+
+type NonceTrackerClient interface {
+	ConfiguredChainID() *big.Int
+	PendingSequenceAt(context.Context, common.Address) (evmtypes.Nonce, error)
+}
+
+type nonceTracker struct {
+	lggr             logger.SugaredLogger
+	nextSequenceMap  map[common.Address]evmtypes.Nonce
+	txStore          NonceTrackerTxStore
+	chainID          *big.Int
+	client           NonceTrackerClient
+	enabledAddresses []common.Address
+
+	sequenceLock sync.RWMutex
```

Comment:
out of curiosity, does the nonceTracker need a lock? If it's always been called from the broadcaster which has its own lock, then it might be protected already and adding the extra lock just slows down the code. But I'd need to look more carefully to confirm this. It might be safer just to have it unless it's crystal clear that the the calling code is protected by a broadcaster lock.

Reply:
This lock was actually moved over from the Broadcaster so we aren't double locking here.

Reply:
Theoretically, the Broadcaster can process multiple addresses at the same time, which are handled by different go routines (`monitorTxs`) so there is a scenario where SequenceTracker requires locking. Good observation though, I like that we're pushing to clean up unnecessary things 🙌 .